File: devicemonitor/scripts/DeviceMonitor.js
// Configuration
const GITHUB_TOKEN = "" // Add your PAT when deploying
const GITHUB_REPO = "stirlo/thelaboratory.cc"
const GITHUB_PATH = "devicemonitor/data/devices.json"

// Device Configuration - Set this for each device
const DEVICE_NAME = "MiniProductRed13" // Set unique device name here
const DEVICE_MODEL = "iPhone 13 mini"   // Set specific model here

// Thresholds and Intervals
const BATTERY_LOW = 40
const BATTERY_CRITICAL = 20
const BATTERY_READY = 75
const BATTERY_FULL = 100
const STORAGE_WARNING = 90
const CHECK_INTERVAL_CHARGING = 5 * 60 // 5 minutes in seconds
const CHECK_INTERVAL_BATTERY = 20 * 60 // 20 minutes in seconds

function log(type, message) {
    if (['INFO', 'WARN', 'ERROR', 'SUCCESS'].includes(type.toUpperCase())) {
        const timestamp = new Date().toISOString()
        console.log(`${timestamp}: [${type}] ${message}`)
    }
}

async function sendNotification(title, body, critical = false) {
    try {
        log("INFO", `Scheduling notification: ${title} - ${body}`)
        let notification = new Notification()
        notification.title = title
        notification.body = body
        notification.sound = critical ? "alert" : "default"
        notification.threadIdentifier = "DeviceMonitor"

        if (title.includes("battery.0") || title.includes("battery.25")) {
            notification.addAction("battery.circle Settings", "prefs:root=BATTERY_USAGE")
        }
        if (title.includes("externaldrive")) {
            notification.addAction("folder.circle Settings", "prefs:root=STORAGE")
        }

        // Set next check interval based on charging state
        notification.trigger = {
            timeInterval: Device.isCharging() ? CHECK_INTERVAL_CHARGING : CHECK_INTERVAL_BATTERY,
            repeats: false
        }

        await notification.schedule()
        log("SUCCESS", "Notification scheduled successfully")
    } catch (error) {
        log("ERROR", `Failed to schedule notification: ${error}`)
    }
}

async function getGithubFile() {
    if (!GITHUB_TOKEN) {
        throw new Error("GitHub token not configured")
    }

    const url = `https://api.github.com/repos/${GITHUB_REPO}/contents/${GITHUB_PATH}`
    const req = new Request(url)
    req.headers = {
        'Authorization': `token ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3+json'
    }

    try {
        const response = await req.loadJSON()

        if (response.message === "Not Found" || !response) {
            return {
                data: {
                    devices: {
                        macs: {},
                        ipads: {},
                        iphones: {}
                    },
                    last_updated: Date.now()
                },
                sha: null
            }
        }

        if (response.content) {
            const cleanContent = response.content.replace(/\n/g, '')
            const decodedContent = atob(cleanContent)
            const parsedData = JSON.parse(decodedContent)

            return {
                data: parsedData,
                sha: response.sha
            }
        }

        return {
            data: {
                devices: {
                    macs: {},
                    ipads: {},
                    iphones: {}
                },
                last_updated: Date.now()
            },
            sha: null
        }
    } catch (error) {
        log("ERROR", `GitHub fetch error: ${error}`)
        throw error
    }
}

async function updateGithubFile(content, sha) {
    if (!GITHUB_TOKEN) {
        throw new Error("GitHub token not configured")
    }

    const url = `https://api.github.com/repos/${GITHUB_REPO}/contents/${GITHUB_PATH}`
    const req = new Request(url)
    req.method = 'PUT'
    req.headers = {
        'Authorization': `token ${GITHUB_TOKEN}`,
        'Accept': 'application/vnd.github.v3+json',
        'Content-Type': 'application/json'
    }

    try {
        const stringContent = JSON.stringify(content, null, 2)
        const body = {
            message: `Update device status: ${DEVICE_NAME}`,
            content: btoa(stringContent)
        }

        if (sha) {
            body.sha = sha
        }

        req.body = JSON.stringify(body)
        const response = await req.loadJSON()

        return response.content ? true : false
    } catch (error) {
        log("ERROR", `GitHub update error: ${error}`)
        throw error
    }
}

async function checkAllDevicesAndNotify(existingData) {
    try {
        log("INFO", "Starting notification check")
        const allDevices = [
            ...Object.values(existingData?.devices?.iphones || {}),
            ...Object.values(existingData?.devices?.ipads || {}),
            ...Object.values(existingData?.devices?.macs || {})
        ]

        for (const device of allDevices) {
            const battery = device.system_info?.battery
            if (!battery) continue

            log("INFO", `Checking ${device.device_name}: ${battery.percentage}%, charging: ${battery.charging}`)

            if (!battery.charging) {
                if (battery.percentage <= BATTERY_CRITICAL) {
                    await sendNotification(
                        "battery.0 Critical Battery",
                        `${device.device_name} battery critically low at ${battery.percentage}%!`,
                        true
                    )
                } else if (battery.percentage <= BATTERY_LOW) {
                    await sendNotification(
                        "battery.25 Low Battery",
                        `${device.device_name} battery at ${battery.percentage}%`
                    )
                }
            }

            if (battery.charging) {
                if (battery.percentage >= BATTERY_FULL) {
                    await sendNotification(
                        "battery.100.bolt Fully Charged",
                        `${device.device_name} is fully charged`
                    )
                } else if (battery.percentage >= BATTERY_READY) {
                    await sendNotification(
                        "battery.75.bolt Ready to Unplug",
                        `${device.device_name} is charged enough (${battery.percentage}%)`
                    )
                }
            }

            if (device.system_info?.storage?.boot_drive >= STORAGE_WARNING) {
                await sendNotification(
                    "externaldrive.badge.exclamationmark Storage Alert",
                    `${device.device_name} storage is ${device.system_info.storage.boot_drive}% full`
                )
            }
        }
    } catch (error) {
        log("ERROR", `Notification check failed: ${error}`)
    }
}

async function updateDeviceStatus() {
    try {
        const { data: existingData, sha } = await getGithubFile()

        await checkAllDevicesAndNotify(existingData)

        const deviceInfo = {
            device_name: DEVICE_NAME,
            device_type: Device.isPad() ? "ipad" : "iphone",
            device_model: DEVICE_MODEL,
            system_info: {
                os_version: Device.systemVersion(),
                model: DEVICE_MODEL,
                battery: {
                    percentage: Math.round(Device.batteryLevel() * 100),
                    charging: Device.isCharging()
                },
                storage: {
                    boot_drive: 0
                }
            },
            last_updated: new Date().toISOString()
        }

        const updatedData = {
            devices: {
                macs: { ...(existingData?.devices?.macs || {}) },
                ipads: { ...(existingData?.devices?.ipads || {}) },
                iphones: { ...(existingData?.devices?.iphones || {}) }
            },
            last_updated: Date.now()
        }

        const deviceType = Device.isPad() ? "ipads" : "iphones"
        updatedData.devices[deviceType][DEVICE_NAME] = deviceInfo

        const success = await updateGithubFile(updatedData, sha)
        if (!success) {
            throw new Error("Failed to update GitHub")
        }

        log("SUCCESS", `Update completed successfully for ${DEVICE_NAME}`)

    } catch (error) {
        log("ERROR", `Update failed: ${error}`)
        throw error
    }
}

// Run update and schedule next notification
await updateDeviceStatus()

File: devicemonitor/scripts/devicemonitor.py
#!/usr/bin/env python3

import urllib.request
import json
import os
from datetime import datetime, UTC
import base64
import sys
import subprocess

# Configuration
GITHUB_TOKEN = os.getenv('GITHUB_TOKEN')
if not GITHUB_TOKEN:
    with open("/Users/DT/Library/Logs/devicemonitor.log", 'a') as f:
        f.write(f"{datetime.now(UTC)}: Error - GITHUB_TOKEN not found in environment\n")
    sys.exit(1)

GITHUB_REPO = "stirlo/thelaboratory.cc"
GITHUB_PATH = "devicemonitor/data/devices.json"
TEMP_FILE = "/tmp/devices.json"
LOG_FILE = "/Users/DT/Library/Logs/devicemonitor.log"

# Battery thresholds
BATTERY_CRITICAL = 20
BATTERY_LOW = 40
BATTERY_HIGH = 75
BATTERY_FULL = 100

def notify(title, message):
    """Send macOS notification"""
    os.system(f'osascript -e \'display notification "{message}" with title "{title}"\'')

def make_github_request(url, method='GET', data=None):
    """Handle GitHub API requests"""
    headers = {
        "Authorization": f"token {GITHUB_TOKEN}",
        "Accept": "application/vnd.github.v3+json",
        "Content-Type": "application/json"
    }

    request = urllib.request.Request(
        url,
        data=json.dumps(data).encode('utf-8') if data else None,
        headers=headers,
        method=method
    )

    try:
        with urllib.request.urlopen(request) as response:
            return json.loads(response.read().decode('utf-8'))
    except Exception as e:
        with open(LOG_FILE, 'a') as f:
            f.write(f"{datetime.now(UTC)}: API Error - {str(e)}\n")
        raise

def get_mac_info():
    """Gather Mac system information"""
    name = subprocess.getoutput('scutil --get ComputerName')
    os_ver = subprocess.getoutput('sw_vers -productVersion')
    build_num = subprocess.getoutput('sw_vers -buildVersion')

    try:
        temp = float(subprocess.getoutput('osx-cpu-temp').split()[0])
    except:
        temp = 0

    battery_info = subprocess.getoutput('pmset -g batt')

    if "Smart-UPS" in battery_info:
        try:
            battery_pct = int(next(x for x in battery_info.split() if '%' in x).rstrip('%'))
            charging = "AC attached" in battery_info
        except:
            battery_pct = 100
            charging = True
    else:
        battery_pct = 100
        charging = True

    try:
        boot_used = int(subprocess.getoutput("df -h / | awk 'NR==2 {print $5}' | sed 's/%//'"))
    except:
        boot_used = 0

    return {
        "device_name": name,
        "device_type": "mac",
        "system_info": {
            "os_version": os_ver,
            "build_number": build_num,
            "temperature": temp,
            "battery": {
                "percentage": battery_pct,
                "charging": charging
            },
            "storage": {
                "boot_drive": boot_used
            }
        },
        "last_updated": datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
    }

def update_github():
    """Update device info on GitHub"""
    try:
        # Get current file
        api_url = f"https://api.github.com/repos/{GITHUB_REPO}/contents/{GITHUB_PATH}"
        current_file = make_github_request(api_url)

        # Get current content
        with urllib.request.urlopen(current_file['download_url']) as response:
            current_data = json.loads(response.read().decode())

        # Update with new data
        mac_info = get_mac_info()
        device_name = mac_info['device_name']

        if 'macs' not in current_data['devices']:
            current_data['devices']['macs'] = {}

        current_data['devices']['macs'][device_name] = mac_info
        current_data['last_updated'] = datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")

        # Handle notifications
        battery_level = mac_info['system_info']['battery']['percentage']
        charging = mac_info['system_info']['battery']['charging']

        battery_info = subprocess.getoutput('pmset -g batt')
        if "Smart-UPS" in battery_info:
            if not charging:
                notify("🚨 UPS ON BATTERY!", f"{device_name} UPS is running on battery power")
        else:
            if not charging and battery_level <= BATTERY_CRITICAL:
                notify("🚨 OMG CHARGE NOWWW!", f"{device_name} at {battery_level}%")
            elif not charging and battery_level <= BATTERY_LOW:
                notify("Battery Getting Low", f"{device_name} at {battery_level}%")

        # Push update
        update_data = {
            "message": f"Update {device_name} status",
            "content": base64.b64encode(json.dumps(current_data, indent=2).encode()).decode(),
            "sha": current_file['sha']
        }

        make_github_request(api_url, method='PUT', data=update_data)

        with open(LOG_FILE, 'a') as f:
            f.write(f"{datetime.now(UTC)}: Successfully updated device status\n")

    except Exception as e:
        with open(LOG_FILE, 'a') as f:
            f.write(f"{datetime.now(UTC)}: Error updating status - {str(e)}\n")

if __name__ == "__main__":
    update_github()
